### PR TITLE
Add Upsample wrapper and DRAM Slicing support to TT-CNN

### PR DESCRIPTION
**Pull request by ianastasijevicTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/29509)

### Problem description
We want to migrate Panoptic-DeepLab to TT-CNN to use a set of common wrappers instead of rolling our own. At the moment, however, we rely on those wrappers to handle DRAM slicing for convolutions, max pool and upsample (we are very L1 constrained).

This PR adds Upsample and DRAM slicing support to TT-CNN so that we may migrate.

### What's changed
- New wrapper for Upsample
- DRAM slicing support for Conv2D (width, height, channel sliced), MaxPool2D (channel sliced), Upsample (channel sliced)
- Unit test coverage

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_cnn/tt/builder.py [hunk 1]
```diff
@@ -353,6 +483,15 @@ def to_compute_config(configuration: Conv2dConfiguration, device: ttnn.Device):
...
+def to_slice_config(configuration: Conv2dConfiguration):
+    if configuration.slice_strategy.get_slice_type() is None:
+        return None
+    return ttnn.Conv2dSliceConfig(
+        slice_type=configuration.slice_strategy.get_slice_type(),
+        num_slices=configuration.slice_strategy.get_num_slices(),
+    )
+
+
 class TtConv2d:
```
Comment (esmalTT) [inline models/tt_cnn/tt/builder.py:489]: I wonder if we should treat `TtConv2d` (and other wrappers) as closed and instead have a higher order wrapper that could compose TtConv2ds into a sliced convolution.

The advantage would be that it decouples the functionality, making it easier to test/maintain.

File: models/tt_cnn/tt/builder.py [hunk 2]
```diff
@@ -11,7 +11,61 @@
...
+    def get_num_slices(self):
+        return self.num_slices
+
+
+# Currently, channel slicing is not natively supported by the Conv2D operation and must be done manually
+# For now, we identify channel slicing by having a num_slices > 0 with a slice_type of None
+# Normally, num_slices of 0 for a not None slice_type means auto-slice, so we don't have that feature for channel slice
+# Once Conv2D supports channel slicing natively, we should be able to support it without API changes
+@dataclass
+class NoSliceStrategyConfiguration(SliceStrategyConfiguration):
```
Comment (esmalTT) [inline models/tt_cnn/tt/builder.py]: Do we need this? Slice strategy could just be  
```
Union[
    HeightSliceStrategyConfiguration,
    WidthSliceStrategyConfiguration,
    ChannelSliceStrategyConfiguration,
]
```

And just make the member variables Optional[SliceStrategyConfiguration]

File: models/tt_cnn/tt/builder.py [hunk 3]
```diff
@@ -105,6 +159,7 @@ class Conv2dConfiguration:
     output_layout: ttnn.Layout = ttnn.TILE_LAYOUT
 
     sharding_strategy: ShardedStrategyConfiguration = AutoShardedStrategyConfiguration()
+    slice_strategy: SliceStrategyConfiguration = NoSliceStrategyConfiguration()
```
Comment (esmalTT) [inline models/tt_cnn/tt/builder.py]: Probably makes more sense to make this `Optional[SliceStrategyConfiguration]`

**Review by esmalTT — COMMENTED** [review]
Thanks for the contribution! 

I have some high level feedback for now, lets discuss and then I can come back for a more detailed look. 

My big question is can we somehow leave the layer wrappers (TtConv2d) closed and just compose them to create the slicing behaviour?

File: models/tt_cnn/tt/builder.py (hunk 1 above)
Comment (ianastasijevicTT) [reply models/tt_cnn/tt/builder.py:489]: My thinking is that eventually channel slicing might be supported by convolution and the other ops as a built-in feature passed as any other parameter. Once that happens there won't be much point IMO in the higher order wrapper, and we can simply remove the clunky python implementation at that stage without breaking the API (since it's encapsulated in the wrapper and invisible externally).

If we make a separate wrapper for sliced operations, we'll need a different one for conv and upsample/pool as they handle channel slicing differently, which to me sounds like it might increase maintenance burden, or a single wrapper function with different logic based on the type of operation being sliced.

As an intermediate solution I can separate out the channel slicing logic into their own class functions within the wrappers. This won't functionally change anything, but it will keep the __call__ function short and it might be easier to remove later if/when channel slicing is properly supported.

Let me know what your preferred way of handling this issue would be.

Comment (esmalTT) [reply models/tt_cnn/tt/builder.py:489]: Good point about deprecating our own explicit slicing in the future.  

I like your plan, let's do it!

Comment (ianastasijevicTT) [reply models/tt_cnn/tt/builder.py:489]: Channel slicing moved to own method.

File: models/tt_cnn/tt/builder.py (hunk 2 above)
Comment (ianastasijevicTT) [reply models/tt_cnn/tt/builder.py]: Refactored.

File: models/tt_cnn/tt/builder.py [hunk 4]
```diff
@@ -294,7 +354,68 @@ def from_torch(
...
 
-LayerConfiguration = Union[Conv2dConfiguration, MaxPool2dConfiguration]
+@dataclass(frozen=True)
+class UpsampleConfiguration:
+    input_height: int
+    input_width: int
+    channels: int
+    batch_size: int
+    scale_factor: Union[int, Tuple[int, int]]
+    mode: str = "nearest"
```
Comment (pavlejosipovic) [inline models/tt_cnn/tt/builder.py:365]: we use some enum style thing here (not a string)

Comment (ianastasijevicTT) [reply models/tt_cnn/tt/builder.py:365]: Can you elaborate on this? I can't find where it's defined. Upsample pybind takes a string-type mode parameter and it's passed as a string all the way to create_program where it's checked for "nearest" and "bilinear" and the appropriate program factory is instantiated.

**Comment by ianastasijevicTT** [comment]
All post commit green https://github.com/tenstorrent/tt-metal/actions/runs/18344464036
Blackhole post commit red (same error as latest main) https://github.com/tenstorrent/tt-metal/actions/runs/18348489023
Model perf green https://github.com/tenstorrent/tt-metal/actions/runs/18344538365
Device perf green https://github.com/tenstorrent/tt-metal/actions/runs/18344519114
Frequent models green https://github.com/tenstorrent/tt-metal/actions/runs/18344559552

File: models/tt_cnn/tt/builder.py [hunk 5]
```diff
@@ -353,6 +474,18 @@ def to_compute_config(configuration: Conv2dConfiguration, device: ttnn.Device):
     )
 
 
+def to_slice_config(configuration: Conv2dConfiguration):
```
Comment (esmalTT) [inline models/tt_cnn/tt/builder.py]: This thing could probably take a slice_config and not a `Conv2dConfiguration`.

File: models/tt_cnn/tt/builder.py [hunk 6]
```diff
@@ -362,12 +495,56 @@ def __init__(
...
+                        [
+                            configuration.out_channels,
+                            end_idx,
+                            configuration.kernel_size[0],
+                            configuration.kernel_size[1],
+                        ],
+                    )
+                    self.weight_slices.append(weight_slice)
+            else:
+                # Weights are on host - convert to torch, slice, then convert back to TTNN
```
Comment (esmalTT) [inline models/tt_cnn/tt/builder.py:534]: Not something we can easily address in the PR, but ideally we don't want to go back to `torch` and back here. I'm guessing `ttnn.slice` does not work on host? 

Can you open an issue to do the weight preprocessing using TTNN so someone can address this in the future?

**Review by esmalTT — APPROVED** [review]
LGTM - nice work!

File: models/tt_cnn/tt/builder.py (hunk 5 above)
Comment (ianastasijevicTT) [reply models/tt_cnn/tt/builder.py]: Refactored to take Optional[SliceStrategy] instead of Conv2dConfiguration. Changed all uses to pass configuration.slice_strategy instead of the whole configuration.

File: models/tt_cnn/tt/builder.py (hunk 6 above)
Comment (ianastasijevicTT) [reply models/tt_cnn/tt/builder.py:534]: Discussed offline - will open issue for TTNN slicing on host.
